x-pack/plugins/apm/server/lib/transaction_groups/fetcher.ts

```diff
@@ -36,66 +38,159 @@ interface TopTraceOptions {
-  }
+  delete projection.body.aggs;
+
+  const size = bucketSize + 1;
 
-  const params = mergeProjection(projection, {
+  const request = mergeProjection(projection, {
+    size: 0,
     body: {
-      size: 0,
-      query: {
-        bool: {
-          // prefer sampled transactions
-          should: [{ term: { [TRANSACTION_SAMPLED]: true } }],
-        },
-      },
       aggs: {
         transaction_groups: {
-          composite: {
-            size: bucketSize + 1, // 1 extra bucket is added to check whether the total number of buckets exceed the specified bucket size.
-            sources: [
-              ...(isTopTraces
-                ? [{ service: { terms: { field: SERVICE_NAME } } }]
-                : []),
-              { transaction: { terms: { field: TRANSACTION_NAME } } },
-            ],
-          },
-          aggs: {
-            sample: { top_hits: { size: 1, sort } },
-            avg: { avg: { field: TRANSACTION_DURATION } },
-            p95: {
-              percentiles: {
-                field: TRANSACTION_DURATION,
-                percents: [95],
-                hdr: { number_of_significant_value_digits: 2 },
-              },
-            },
-            sum: { sum: { field: TRANSACTION_DURATION } },
-          },
+          ...(isTopTraces
+            ? {
+                composite: {
+                  sources: [
+                    { [SERVICE_NAME]: { terms: { field: SERVICE_NAME } } },
+                    {
+                      [TRANSACTION_NAME]: {
+                        terms: { field: TRANSACTION_NAME },
+                      },
+                    },
+                  ],
+                  size,
+                },
+              }
+            : {
+                terms: {
+                  field: TRANSACTION_NAME,
+                  size,
+                },
+              }),
         },
       },
     },
   });
 
-  return client.search<Transaction, typeof params>(params);
+  const params = {
+    request,
+    setup,
+  };
+
+  const metrics = await Promise.all([
+    getSamples(params),
+    getAvg(params),
+    getSum(params),
+    !isTopTraces ? getPercentiles(params) : Promise.resolve(undefined),
+  ]);
+
+  const items: TransactionGroupData[] = [];
+
+  type Metric = ValuesType<Exclude<ValuesType<typeof metrics>, undefined>>;
+
+  const allMetrics = flatten(metrics as any).filter(Boolean) as Metric[];
+
+  allMetrics.forEach((metric) => {
+    // we use indexOf so we can replace the existing item with a new one
+    // this will give us type safety (Object.assign is unsafe)
+    let indexOf = findIndex(items, (i) => isEqual(i.key, metric.key));
+    let item = items[indexOf];
+
+    if (indexOf === -1) {
+      const newItem = {
+        key: metric.key,
+      };
+      items.push(newItem);
+      item = newItem;
+      indexOf = items.length - 1;
+    }
+
+    const newItem = {
+      ...item,
+      ...metric,
+    };
+
+    items[indexOf] = newItem;
+  });
```

Comment:
This looks pretty intimidating. Can we extract this logic and add some comments?
Is the new logic or what does it replace?

Reply:
I hate it too! I had Object.assign({}) first which was cleaner, but I ran into a bug because there was no type checking of the values I was assigning.

Comments might clear it up but I'm open to suggestions as well. This code is trying to flatten buckets from multiple requests by either a grouping record (`service.name` and `transaction.name`) or a grouping key (`transaction.name`).